### ISSUE TITLE
Workers and parallelism question

### ISSUE THREAD

**Issue by bfulop** [wallabyjs/public#985]
I'm still tweaking my [live CSS/HTML testing](https://github.com/bfulop/interactive_frontend_dev) setup, and noticed that since wallaby runs the tests in parallel, my [WebdriverIO setup](https://github.com/bfulop/interactive_frontend_dev/blob/master/wallaby-wdio.js) runs two times (on a dual core machine). This, in my case opens the browser windows two times ([`wdioclient.init()`](https://github.com/bfulop/interactive_frontend_dev/blob/master/wallaby-wdio.js#L27)).

I've managed to solve this by reducing the workers' numbers to 1:

```
    workers: {
      initial: 1,
      regular: 1
    },
```

This seems to work well, and I have no performance issues.

The webdriverIO library I'm using actually passes commands using a "JSON wire protocol" to a Java service running locally called Selenium Standalone. I'm opening two (or three) browser windows at different sizes (mobile, tablet, desktop) and all the commands in WebdriverIO are async functions (Promises). 

I will ask the maintainers of WebdriverIO if there's a way to get if there are existing Selenium sessions open and reuse those (in the `.init()` call above) instead of opening a new one.

I'm also wondering if I could do something more appropriate in Wallaby? For example if there's a way to generate a worker ID, then I could use that, to target different instances of WebdriverIO. Not sure if this is a good idea though. 

If you have any opinions on this, I would interested in reading it! :)

**Comment by ArtemGovorov** [comment]
> I've managed to solve this by reducing the workers' numbers to 1:

Yes, by default [wallaby.js is trying to utilise multiple cores](https://wallabyjs.com/docs/config/workers.html) of the machine,

>  For example if there's a way to generate a worker ID, then I could use that, to target different instances of WebdriverIO.

We do [pass a unique worker ID](https://wallabyjs.com/docs/config/bootstrap.html) (see `workerId`) to the `setup` function, so if you need to use the ID, it's possible.

**Comment by ArtemGovorov** [comment]
> If you have any opinions on this, I would interested in reading it! :)

Your setup looks pretty impressive! I don't have much experience with WebdriverIO, but I'm more than happy to help with any wallaby related stuff.

**Comment by bfulop** [comment]
Thanks Artem!

I will create a simple repo once I'm beyond the proof of concept phase and you can include it on your site if you want. I really hope that this setup can take away some of the pain of working with CSS and responsive design. 

In the meantime, I got confirmation from the WebdriverIO team, that it is not possible to reuse sessions. So I'll give a shot at using `workerID`...

**Comment by bfulop** [comment]
I just realised, that it might not be possible to do what I wanted, so I prefer to verify it with you.
I have managed to use the workers to create different instances of WebdriverIO and thus open several browser windows at different sizes.

There are two questions that came up:

**A.**  Is it possible to trigger a rerun of a test in both workers if I change my source file? (To test both in mobile and desktop in parallel) (for now, I `require` directly the source file and pass it to webdriverio, that renders html in the browser)

**B.** What happens if I have 5 test files and two workers running, where each worker has its own associated browser size?

So far, both my desktop and mobile related tests are in the same spec file. Should I split them into two separate files (desktop specs and mobile specs) to solve point B. ?

I'm going through different scenarios right now, to see if these points can be resolved...  

The easiest would be in the point of view of parallelism is to do the webdriverio init (where it opens the browser window etc) in a `before` hook. But this part is very slow, so better to move it to Wallaby startup. 

 (will soon leave the office and will only be able to reply or continue tomorrow)

**Comment by bfulop** [comment]
I think I will just stick with one worker and save myself a lot of complications. The performance is great and the results are predictable.

Also, I've noticed that even if I set `workers.regular` to `4`, and log out `wallaby.workerId` in the `setup` function, at first I get `1`, `0`, then `2`. Later, when I edit my source files, or my tests, I only get `0` in the console, the other workers never seem to run. What are the conditions when Wallaby uses the other workers?

**Comment by ArtemGovorov** [comment]
> Is it possible to trigger a rerun of a test in both workers if I change my source file

> What happens if I have 5 test files and two workers running

> What are the conditions when Wallaby uses the other workers?

Wallaby dynamically calculates what tests it needs to run based on your changes and distributes the work amongst the workers. Initially, it may run 2 test files in one worker and 3 test files in the other. Later on, when you're only editing one file, wallaby just allocates the first worker (id = `0`) to do the work.

**Comment by bfulop** [comment]
Now that I know better how Wallaby uses workers, I realised, that in fact I don't really need to mess with the them at all...

When you have three tests, it's a bit scary to see that now you have two more browser windows popping up, but it will surely make sense when I'll have 20 or 50 tests running. The number of browser windows is not linear to the number of tests you have, but the hardware, etc. 

Also, I'm now adding the worker id to the title of the browser window so it's easy to find the one with id `0` to see the live feedback of my editing.

So I just removed workers setting completely and trust Wallaby to handle the optimal numbers.
